modules/tinymce/src/core/main/ts/mode/Readonly.ts

```diff
@@ -38,23 +39,54 @@ const registerReadOnlyInputBlockers = (editor: Editor): void => {
     }
   });
 
-  editor.on('input', (e) => {
-    if (!e.isComposing && isReadOnly(editor)) {
-      const undoLevel = editor.undoManager.add();
-      if (Type.isNonNullable(undoLevel)) {
-        editor.undoManager.undo();
+  // Prevent undo level additions during composition
+  editor.on('input compositionend', (e) => {
+    if (isReadOnly(editor)) {
+      // Skip undo level addition during composition
+      if (!e.isComposing) {
+        const undoLevel = editor.undoManager.add();
+        if (Type.isNonNullable(undoLevel)) {
+          editor.undoManager.undo();
+        }
       }
     }
   });
 
-  editor.on('compositionend', () => {
+  // Set up mutation observer to detect and revert unintended changes
+  const observer = new MutationObserver((mutations) => {
     if (isReadOnly(editor)) {
-      const undoLevel = editor.undoManager.add();
-      if (Type.isNonNullable(undoLevel)) {
-        editor.undoManager.undo();
+      let needsRevert = false;
+
+      mutations.forEach((mutation) => {
+        if (mutation.type === 'characterData' || mutation.type === 'childList') {
+          // Mark that we need to revert
+          needsRevert = true;
+        }
+      });
+
+      // Only create one undo level for all mutations in this batch
+      if (needsRevert) {
+        const undoLevel = editor.undoManager.add();
+        if (Type.isNonNullable(undoLevel)) {
+          editor.undoManager.undo();
+        }
       }
     }
   });
+
+  editor.on('compositionstart', () => {
+    if (isReadOnly(editor)) {
+      observer.observe(editor.getBody(), {
+        characterData: true,
+        childList: true,
+        subtree: true
+      });
+    }
+  });
+
+  editor.on('compositionend', () => {
+    observer.disconnect();
```

Comment:
I think it's better to do the filtering of the mutations here by calling the `takeRecords()` that call always return any pending mutations. The callback is not always executed for every mutation it has a async nature to it. Also I think that is a bit cleaner.